src/ir_passes/bundle_elim.rs

```diff
@@ -0,0 +1,233 @@
+        let end = comp.get(range.end).clone();
+
+        let len = len.as_concrete(comp).unwrap();
+
+        // if we need to preserve external interface information, we can't have bundle ports in the signature.
+        if comp.src_info.is_some()
+            && matches!(owner, crate::ir::PortOwner::Sig { .. })
+        {
+            assert!(
+                len == 1,
+                "Bundle ports in the signature are not supported."
+            );
+
+            // need to preserve the original portidx here to save the source information.
+            return vec![pidx];
+        }
+
+        let ports = (0..len)
+            .map(|i| {
+                let binding: HashMap<_, _> = [(idx, i)].into();
+
+                let offset = start.offset.resolve(comp, &binding);
+                let start = comp.add(Time {
+                    event: start.event,
+                    offset,
+                });
+
+                let offset = end.offset.resolve(comp, &binding);
+                let end = comp.add(Time {
+                    event: end.event,
+                    offset,
+                });
+                let live = Liveness {
+                    idx,
+                    len: one,
+                    range: Range { start, end },
+                };
+
+                let owner = match &owner {
+                    crate::ir::PortOwner::Sig { dir } => {
+                        crate::ir::PortOwner::Sig { dir: dir.clone() }
+                    }
+                    crate::ir::PortOwner::Inv { inv, dir, base } => {
+                        crate::ir::PortOwner::Inv {
+                            inv: *inv,
+                            dir: dir.clone(),
+                            base: Foreign::new(
+                                self.context
+                                    .get(base.owner)
+                                    .get(&base.key)
+                                    .unwrap()[i as usize],
+                                base.owner,
+                            ),
+                        }
+                    }
+
+                    crate::ir::PortOwner::Local => crate::ir::PortOwner::Local,
+                };
+
+                comp.add(Port {
+                    live,
+                    owner,
+                    info,
+                    width,
+                })
+            })
+            .collect();
+        // delete the original port
+        comp.delete(pidx);
+        // delete the corresponding parameter
+        comp.delete(idx);
+        ports
+    }
+
+    /// Compiles the signature of a component and adds the new ports to the context mapping.
+    fn sig(&self, comp: &mut Component) -> HashMap<PortIdx, Vec<PortIdx>> {
+        // need to collect here because of ownership issues.
+        comp.ports()
+            .idx_iter()
+            .filter(|p| comp.get(*p).is_sig())
+            .collect_vec() // collect here to avoid ownership issues
```

Comment:
What ownership issue occurs if you don't do this? Is it because `port` takes mutable ownership of the component?